worker/metrics/collect/manifold.go

```diff
@@ -151,10 +160,28 @@ func newCollect(config ManifoldConfig, context dependency.Context) (*collect, er
 		unitTag: unitTag.String(),
 		paths:   paths,
 	}
-	var listener stopper
-	charmURL, validMetrics, err := readCharm(unitTag, paths)
+	var (
+		listener     stopper
+		charmURL     *corecharm.URL
+		validMetrics map[string]corecharm.Metric
+	)
+	err = retry.Call(retry.CallArgs{
+		Func: func() error {
+			var err error
+			charmURL, validMetrics, err = readCharm(unitTag, paths)
+			return err
+		},
+		IsFatalError: func(err error) bool {
+			return errors.Cause(err) != errCharmDirNotExist
+		},
+		Attempts:    10,
+		Delay:       time.Second,
+		MaxDelay:    time.Duration(30) * time.Second,
```

Comment:
It's a little arbitrary, but the idea is to wait a few minutes and then give up -- you might be on a heavily thrashed machine (a certain I/O-starved OpenStack deployment I know of comes to mind!) but at some point, if the charm dir doesn't show, it's reasonable to give up.

I was lazy and used the DoubleDelay because it was there. But what I'd really like is range(1..25) which sums to 300, a perfect five minutes of backoff and retry. I should actually do that...